libdnf/dnf-sack.cpp

```diff
@@ -230,12 +230,20 @@ can_use_repomd_cache(FILE *fp_solv, unsigned char cs_repomd[CHKSUM_BYTES])
 {
     unsigned char cs_cache[CHKSUM_BYTES];
 
-    if (fp_solv &&
-        !checksum_read(cs_cache, fp_solv) &&
-        !checksum_cmp(cs_cache, cs_repomd))
-        return 1;
+    if (!fp_solv ||
+        checksum_read(cs_cache, fp_solv) ||
+        checksum_cmp(cs_cache, cs_repomd))
+        return 0;
 
-    return 0;
+    char solvfile_version_cache[SOLVFILE_VERSION_BYTES];
+
+    if (!fp_solv ||
+        solvfile_version_read(solvfile_version_cache, fp_solv) ||
+        g_strcmp0(solvfile_version_cache, solv_toolversion)) {
```

Comment:
The issue with this is that if you read the current .solv file format that doesn't have the version at this position, you're reading random binary data. While the probability is low, you can hit the exact string representing the version and bad things will happen.

Do we simply accept the risk (possibly figuring out what exactly will happen, probably the same thing that was reported already for the mismatching cached data), or do we try to make this bulletproof?

I'm thinking whether we could change the format of the .solv files more radically and fix the issue for good. Briefly inspecting the .solv files, each of them seems to start with `"SOLV"` (and a few other bytes that seem to be constant before any variable data). We could base a reliable check on that and introduce a new and better format along the lines of:
```
4 bytes dnf .solv file magic
4 bytes dnf .solv file version
8 bytes LIBSOLV_TOOLVERSION (with some extra padding)
32 bytes repodata checksum
libsolv data
```

This wouldn't be too hard to implement and would be reliable in the sense that if the file starts with SOLV, we'd just drop it and rebuild in the new format. Otherwise we'd check the magic, handling an error as we see fit (at least log it). Then similarly check the two versions that follow it, then check the checksum and finally load the data. I've checked that libsolv doesn't seek on the `FILE *` we pass to it, so it'll start reading from where ever we point.

Optionally we can include the checksum type in the new binary format, as it seems to occur semi-often that we need to change the checksum algorithm. But that can be gracefully handled later via the dnf file version at a later point as well.

Reply:
> The issue with this is that if you read the current .solv file format that doesn't have the version at this position, you're reading random binary data. While the probability is low, you can hit the exact string representing the version and bad things will happen.

It is true but this change also offsets the checksum (meaning its no longer the last 32 bytes in the file) so while reading the current .solv file for the checksum we read 5 random bytes and then 27 bytes of the beginning of the checksum. I believe it is a pretty safe bet that this will not match and the cache will be invalidated.

However I would not be opposed to a more robust and better defined solution like you proposed. Let's discuss it on a meeting.